### PR TITLE
Add Velociraptor loader

### PR DESCRIPTION
**Pull request by Zawadidone**
Ref #113

### CODE REVIEW

**Comment by Schamper** [comment]
What if there are multiple drives? I think it would be beneficial if all of the drives are added.

**Comment by Zawadidone** [comment]
Currently Velociraptor only supports one drive.
https://github.com/Velocidex/velociraptor/blob/master/artifacts/definitions/Windows/KapeFiles/Targets.yaml#L32

Not related to this but Velociraptor has support for different file accessors, but my prefered way is using `ntfs`. The other accessors need a different implementation in the loader that should support all three accessors (`ntfs`, `lazy_ntfs` and `auto`. But that is not part of this issue #115.

https://github.com/Velocidex/velociraptor/blob/master/artifacts/definitions/Windows/KapeFiles/Targets.yaml#L2234

**Comment by Schamper** [comment]
Is it expected that Velociraptor will at some point support multiple disks, and that it will work similarly to this? Otherwise it might be good to add support for that ahead of time?

File: dissect/target/loaders/velociraptor.py [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+from pathlib import Path
+from typing import TYPE_CHECKING
+
+from dissect.target.loaders.dir import DirLoader, find_dirs, map_dirs
+
+if TYPE_CHECKING:
+    from dissect.target import Target
+
+ROOT_FILESYSTEM = "uploads"
```
Comment (Schamper) [inline dissect/target/loaders/velociraptor.py:11]:
```suggestion
FILESYSTEMS_ROOT = "uploads"
```
Seems more fitting?

File: dissect/target/loaders/velociraptor.py [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+    # Generic.Collectors.File (Windows) and Windows.KapeFiles.Targets (Windows) root filesystem is
+    # 'uploads/<file-accessor>/<drive-name>/'
+    os_type, dirs = find_dirs(path.joinpath(ROOT_FILESYSTEM))
+    if os_type in ["linux", "osx"]:
+        return dirs[0]
+
+    # This suppports usage of the ntfs accessor 'uploads/mft/%5C%5C.%5CC%3A' not the accessors lazy_ntfs or auto
+    if not os_type:
+        # Filter out files that start with '.'
+        windows_path = [p for p in path.joinpath(ROOT_FILESYSTEM + "/mft").iterdir() if not p.name.startswith(".")][0]
```
Comment (Schamper) [inline dissect/target/loaders/velociraptor.py:26]:
```suggestion
        windows_path = [p for p in path.joinpath(ROOT_FILESYSTEM).joinpath("/mft").iterdir() if not p.name.startswith(".")][0]
```
Better to individually join the `mft` path.

Comment (Schamper) [reply dissect/target/loaders/velociraptor.py:26]: Also, what if the `mft` directory doesn't exist?

File: dissect/target/loaders/velociraptor.py [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+        os_type, dirs = find_dirs(windows_path)
+        if os_type == "windows":
+            return dirs[0]
+    return False
```
Comment (Schamper) [inline dissect/target/loaders/velociraptor.py:30]:
```suggestion
    return None
```

File: dissect/target/loaders/velociraptor.py [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+
+from dissect.target.loaders.dir import DirLoader, find_dirs, map_dirs
+
+if TYPE_CHECKING:
+    from dissect.target import Target
+
+ROOT_FILESYSTEM = "uploads"
+
+
+def find_os_directory(path: Path):
```
Comment (Schamper) [inline dissect/target/loaders/velociraptor.py:14]:
```suggestion
def find_os_directory(path: Path) -> Optional[Path]:
``

**Review by Schamper — CHANGES_REQUESTED** [review]

**Comment by Zawadidone** [comment]
> Is it expected that Velociraptor will at some point support multiple disks, and that it will work similarly to this? Otherwise it might be good to add support for that ahead of time?

I don't know I cannot find any issue mentioning this besides https://github.com/Velocidex/velociraptor/issues/1576 which is unrelated. Nonetheless using Velociraptor VQL this could be supported but I have not seen usage of that.

File: dissect/target/loaders/velociraptor.py (hunk 2 above)
Comment (Zawadidone) [reply dissect/target/loaders/velociraptor.py:27]: The function should return `None`

File: dissect/target/loaders/velociraptor.py [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+    # This suppports usage of the ntfs accessor 'uploads/mft/%5C%5C.%5CC%3A' not the accessors lazy_ntfs or auto
+    if not os_type and path.joinpath(FILESYSTEMS_ROOT).joinpath("mft").exists():
+        # Filter out files that start with '.'
+        windows_path = [
+            p for p in path.joinpath(FILESYSTEMS_ROOT).joinpath("mft").iterdir() if not p.name.startswith(".")
+        ][0]
+        os_type, dirs = find_dirs(windows_path)
+        if os_type == "windows":
+            return dirs[0]
+    return None
```
Comment (Schamper) [inline dissect/target/loaders/velociraptor.py:32]:
```suggestion
    fs_root = path.joinpath(FILESYSTEMS_ROOT)
    os_type, dirs = find_dirs(fs_root)
    if os_type in ["linux", "osx"]:
        return dirs[0]

    # This suppports usage of the ntfs accessor 'uploads/mft/%5C%5C.%5CC%3A' not the accessors lazy_ntfs or auto
    mft_root = fs_root.joinpath("mft")
    if not os_type and mft_root.exists():
        # Filter out files that start with '.'
        windows_path = [p for p in mft_root.iterdir() if not p.name.startswith(".")][0]
        os_type, dirs = find_dirs(windows_path)
        if os_type == "windows":
            return dirs[0]

    return None
```
Re-use some path variables to make it a bit easier to read.

**Review by Schamper — CHANGES_REQUESTED** [review]

**Review by Schamper — APPROVED** [review]

**Comment by Zawadidone** [comment]
Thanks!

**Pull request merged**